clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -226,11 +227,26 @@ protected synchronized boolean ensureCoordinatorReady(final Timer timer) {
             return true;
 
         do {
+            if (findCoordinatorException != null && !(findCoordinatorException instanceof RetriableException)) {
+                final RuntimeException fatalException = findCoordinatorException;
+                findCoordinatorException = null;
+                throw fatalException;
+            }
             final RequestFuture<Void> future = lookupCoordinator();
             client.poll(future, timer);
 
             if (!future.isDone()) {
                 // ran out of time
+                future.addListener(new RequestFutureListener<Void>() {
```

Comment:
Do we want this listener only when coordinator lookup is triggered through `ensureCoordinatorReady`? Other paths may use the future from a call to `ensureCoordinatorReady`. Conversely, we may use a future which was sent through another path here. Could we instead move this logic to `lookupCoordinator` so that it is handled consistently?

Reply:
@hachikuji I actually did it intentionally: if we move the logic to the callee then any `client.poll` may trigger the callback and potentially throw the exception other than just `consumer.poll` as we are testing in the unit test. Currently there are three callers of `lookupCoordinator` and AFAIK the other two do not need to check if the future failure is non-retriable and hence need to be thrown. Of course if in the future we add another caller which does want to check then this would be vulnerable.

So I'd propose that if we make this behavior consistent inside `lookupCoordinator`, then we need to mark all public APIs that may leads to it to `@throw` those exceptions (today `consumer.poll` and some other APIs do have the marks, but not all of them). WDYT?

Reply:
Regardless of where the `lookupCoordinator` is triggered, we are only raising it from `ensureCoordinatorReady`, so I am not sure I follow the point about raising from other contexts.

Note there doesn't appear to be any logic preventing multiple listeners from getting attached to the future. I think it would be better to always attach the listener when the future is created.